### ISSUE TITLE
Sass variable conventions

### ISSUE THREAD

**Issue by brockfanning** [open-sdg/open-sdg#712]
Followup from #705 

@SavvasStephanides 
Cool, I'm definitely open to changing things around a particular convention - especially now, before 1.0.0 is released! Let me talk through some stuff though, to get it straight in my head.

Our main goal here is to allow countries to override colors for all the bits and pieces of the platform, independently. So, for example, we want to allow a country to override the footer background color _only_, while leaving everything else the same.

So our convention may be a bit different from a typical web site or design system. For example a typical web site or design system may have variables like `color-primary`, `color-primary-dark`, etc. But we need more granular "component-level" variables.

At the same time, the design-system-ish variables might be useful too. For example, what if a country wants to change the background colors across the site. Having to override the background color of every component seems cumbersome. It would be much better if they could change one variable.

So I'm imagining two separate files of color variables. One file, say `_colors-global.scss`, is the design-system-y approach, which has component-agnostic variables like `color-primary`, etc. (or whatever we decide on as a standard). The other file, say `_colors-by-component.scss` has the granular component-level variables, like `footer-background-color` (or again, whatever we decide on as a standard). **The second file would not contain any actual colors** (hex values, etc) and instead would just have references to the variables in `_colors-global.scss`.

Or maybe instead of two separate files they could just be different sections of `_colors.scss`?

That way, if a country wants to override background colors across the site, they can override one of the "colors-global" variables. And if they instead just want to override the background color of a particular component, they can override a "colors-by-component" variable.

Anyway just brainstorming here. Does that make sense to you?

Regarding the actual standard we use, I'm flexible. I do think they should all include some indication of whether they are a foreground color ("color") or background color ("background-color"). And if they are specific to "hover", "focus", etc, they'll need to include that. And also they'll need something for "high-contrast". And if they're component-specific, that needs to be there too, probably at the beginning.

I like your `element-modifier-extra property` idea, but maybe with the component when needed. So normally `element-modifier-extra property`, but if it's component-specific, then `component-element-modifier-extra property`?

**Comment by SavvasStephanides** [comment]
Thanks for the reply @brockfanning.

> Our main goal here is to allow countries to override colors for all the bits and pieces of the platform, independently. So, for example, we want to allow a country to override the footer background color only, while leaving everything else the same.

That's great! I really think having a standard structure in our variables will benefit our users this way. It will make the variables easy to find and thus easier to override. If for example someone wants to change the background color of the footer they can first look in the list of variables for the element `$footer`, then the modifier `-backgroundColor` and finally any extra details, such as `-highContrast`, `-mobile`, etc... This is slightly similar to the [BEM method](https://css-tricks.com/bem-101/) for naming CSS classes.

> At the same time, the design-system-ish variables might be useful too. For example, what if a country wants to change the background colors across the site. Having to override the background color of every component seems cumbersome. It would be much better if they could change one variable.

That's a good point. Definitely helps if we have some global colours as well. Definitely helps to discuss a naming convention for these types of variables. Two ideas come to mind:

1. Skip the "element" part altogether. So for example if we need to change the link colours globally, we can name the variable `$link-color`. 

2. Include the "element" part and have the whole site as the element. Possibly call it something like `$document-link-color`.

> I like your element-modifier-extra property idea, but maybe with the component when needed. So normally element-modifier-extra property, but if it's component-specific, then component-element-modifier-extra property?

This is basically the BEM method (block-element-modifier) I linked above, but we'll need something extra for high contrast etc.

So the way I see this being structured is this:

1. The global-level attributes will be declared and used at the top:

`_colors_global.scss`
```scss
$document-backgroundColor-highContrast: #000000;
//or
$backgroundColor-highContrast: #000000;
```

`_global.scss`
```scss
body #container{
    background-color: $document-backgroundColor-highContrast;
}
```
I do like your idea of naming the top-level document "global".

2. Then any component-specific colors can be declared further down:

`_colors_footer.scss`
```scss
$footer-backgroundColor-highContrast: #333333;
```

Obviously this is open for discussion. The main idea is that if someone wants to override a specific variable, they can find it easily with their code editor's search tool to be able to know what to override and I think this method will give some clarity.

**Comment by brockfanning** [comment]
@SavvasStephanides I've given this a try in #715. A few notes:
1. I found it made a bit more sense (to me at least) with three categories: top-level, element-level, and component-level.
2. I was intending that the component-level would only have references to the higher-level variables, but in practice that wasn't possible. We just have too many different one-off colors being used.
3. As part of this I tried to flesh out the variables for the status types (complete, notstarted, etc.) and I gave them the full set of normal and high-contrast variables.
4. Similarly, I did the same for goal banners and the footer.
